playbook-base.yaml

```
@@ -1,76 +1,160 @@
+      loop_control:
+        label: "/{{ item.path }}"
+      when: item.state == 'file' and item.path is not regex('\\.jinja2$')
 
-  - name: Copy and render template files
-    ansible.builtin.template:
-      src: "{{ item.src }}"
-      dest: "/{{ item.path | regex_replace('\\.jinja2$', '') }}"
-    with_community.general.filetree: "{{ playbook_dir }}/files/"
-    loop_control:
-      label: "/{{ item.path | regex_replace('\\.jinja2$', '') }}"
-    when: item.state == 'file' and item.path is regex('\\.jinja2$')
+    # Kubelets are configured to _not_ use systemd-resolved.
+    - name: Systemd-resolved is disabled
+      ansible.builtin.systemd:
+        name: systemd-resolved
+        state: stopped
+        enabled: false
+        masked: true
 
-  - name: Copy files
-    ansible.builtin.copy:
-      src: "{{ item.src }}"
-      dest: "/{{ item.path }}"
-    with_community.general.filetree: "{{ playbook_dir }}/files/"
-    loop_control:
-      label: "/{{ item.path }}"
-    when: item.state == 'file' and item.path is not regex('\\.jinja2$')
+    - name: CRI-O is started and enabled
+      ansible.builtin.systemd:
+        name: crio
+        state: started
+        enabled: yes
+    - name: Kubelet is enabled  # Storage nodes override kubelet.service's WantedBy.
+      ansible.builtin.systemd:
+        name: crio
+        enabled: yes
 
-  - name: Systemd-resolved is disabled
-    ansible.builtin.systemd:
-      name: systemd-resolved
-      state: stopped
-      enabled: false
-      masked: true
-  - name: CRI-O is started and enabled
-    ansible.builtin.systemd:
-      name: crio
-      state: started
-      enabled: yes
+- name: Control plane node configuration
+  hosts: control-plane
+  remote_user: root
+  tasks:
+    - name: Copy files
+      ansible.builtin.copy:
+        src: files-controlplane/
+        dest: /
+    - name: CP-specific packages are installed
+      community.general.pacman:
+        name: [ k9s ]
+        state: present
 
-# Storage nodes need disks to be manually decrypted, so kubelet should not be enabled on them.
-- name: Enable kubelet
-  hosts: workers:!storage
+- name: Control plane is set up
+  hosts: control-plane
   remote_user: root
-  strategy: free
   tasks:
-  - name: Kubelet is enabled
-    ansible.builtin.systemd:
-      name: kubelet
-      enabled: yes
+    - name: Check for existing cluster
+      register: admin_conf
+      ansible.builtin.stat:
+        path: /etc/kubernetes/admin.conf
+    - meta: end_host
+      when: admin_conf.stat.exists
+
+    # We are not using kubeadm-managed etcd, but we can get advantage of it to generate certificates.
+    - name: Create config file for generating etcd certs
+      ansible.builtin.copy:
+        dest: /tmp/kubeadm-etcd-certs.yaml
+        content: |
+          apiVersion: kubeadm.k8s.io/v1beta3
+          kind: ClusterConfiguration
+          etcd:
+            local:
+              serverCertSANs:
+                - 10.0.0.9
```

```suggestion
                - "{{ ansible_host }}"
```
Or a variable from facts.

---

See lines 6-9 from this file for explanation. 10.9 is not the host's IP.

---

The, ansible has a reserved group called `[all]`. If you put in the inventory toml something like
```toml
host1
host2
host3

[master]
host1

[workers]
host2
host3

[master:vars]
variable=value_for_masters

[workers:vars]
variable=value_for_workers

[all:vars]
variable=value_for_all  ; this variable will be overridden.
```

---

I gave a thought to this yesterday at 1AM. When I did, there were some files containing these IPs that were not templates... but now there are not, so it might as well be a good idea.

---

Hardcoded IPs are out https://github.com/roobre/fukuno/pull/2/commits/9445360b8923f7a162b26f547a54a89de0ab9f40